src/pipecat/services/websocket_service.py

```
@@ -0,0 +1,125 @@
+            return max_wait
+
+    async def _receive_task_handler(self):
+        """Handles WebSocket message receiving with automatic retry logic."""
+        retry_count = 0
+        MAX_RETRIES = 3
+
+        while True:
+            try:
+                await self._receive_messages()
+                logger.debug(f"{self} connection established successfully")
+                retry_count = 0  # Reset counter on successful message receive
+
+            except asyncio.CancelledError:
+                break
+
+            except Exception as e:
+                retry_count += 1
+                if retry_count >= MAX_RETRIES:
+                    message = f"{self} error receiving messages: {e}"
+                    logger.error(message)
+                    await self.push_error(ErrorFrame(message, fatal=True))
+                    break
+
+                logger.warning(f"{self} connection error, will retry: {e}")
+
+                try:
+                    if await self._reconnect_websocket(retry_count):
+                        retry_count = 0  # Reset counter on successful reconnection
+                    wait_time = self.calculate_wait_time(retry_count)
+                    await asyncio.sleep(wait_time)
+                except Exception as reconnect_error:
+                    logger.error(f"{self} reconnection failed: {reconnect_error}")
+                    continue
+
+    @abstractmethod
+    async def _connect_websocket(self):
+        """Implement service-specific websocket connection logic."""
+        pass
+
+    @abstractmethod
+    async def _disconnect_websocket(self):
+        """Implement service-specific websocket disconnection logic."""
+        pass
+
+    @abstractmethod
+    async def _receive_messages(self):
+        """Implement service-specific message receiving logic."""
+        pass
+
+    @abstractmethod
+    async def push_error(self, error: ErrorFrame):
+        """Implement service-specific error handling."""
+        pass
```

`push_error` is a frunction from `FrameProcessor` so we probably don't really need it here.

---

Which means we should report the error differently. That is, we should probably throw a `WebsocketServiceException` and then do the `push_error()` from the service implementation.

---

This was left over code from an earlier implementation. I forgot to remove it from the base class. Maybe the solution is to just remove it from `WebsocketService`?

---

Mmmmm.... actually, we use this in `_receive_task_handler`... So, intead of throwing an exception we should maybe pass (to the `_receive_task_handler`) a way to report an error. For example:

```
async def _receive_task_handler(self, report_error: Callable[[Frame], Awaitable[None]]):
```

and then:

```
                    await report_error(ErrorFrame(message, fatal=True))
```

---

And

```
        self._receive_task = self.get_event_loop().create_task(self._receive_task_handler(self.push_error))
```

---

I like the latter way better. Much cleaner! Just implemented that.